### DISCUSSION TITLE
How can I record a voice at a custom time?

### DISCUSSION THREAD

**Discussion by xandyxor** [pschatzmann/arduino-audio-tools#309] [Q&A] [answered]
#define RECORD_TIME       (5) //Seconds
#define I2S_CHANNEL_NUM   (1)
#define FLASH_RECORD_SIZE (I2S_CHANNEL_NUM * I2S_SAMPLE_RATE * I2S_SAMPLE_BITS / 8 * RECORD_TIME)

int16_t sBuffer[FLASH_RECORD_SIZE ];

What to do next?

**Comment by pschatzmann** [comment, marked as answer]
You can use millis() which provides the time in milliseconds since the start of the microcontroller.
uint64_t now = millis();
uint64_t start = now + 5000; // in 5 seconds
uint64_t end = now + 15000; // in 15 seconds

Then you just copy the audio data when millis()>start && millis<end;
If you copy to a file you need to add some additional logic to close the file (only once!) when it is done.

If you want to use time in hours and minutes you will need to install an NTP library or RTC module: Google will help you!

**Comment by xandyxor** [comment]
hi
can u help me 
how to copy i2s stream data to float array[]?
I need to put i2s audio data into edge impulse to identify
thx

  #include <Arduino.h>
  #include "AudioTools.h"
  // #include "StarWars30.h"
  #define RECORD_TIME (5) //Seconds
  ...
  #define FLASH_RECORD_SIZE (I2S_CHANNEL_NUM * I2S_SAMPLE_RATE * I2S_SAMPLE_BITS / 8 * RECORD_TIME)
  
  float sBuffer[FLASH_RECORD_SIZE];
  
  I2SStream i2s;  // Output to I2S
  MemoryStream music(sBuffer, sizeof(sBuffer)/sizeof(sBuffer[0]));
  StreamCopyT<int16_t> copier(i2s, music); // copies sound into i2s
  
  ...
  features_signal.get_data = &sBuffer;
  EI_IMPULSE_ERROR res = run_classifier(&features_signal, &result, false /* debug */);
  ...

**Comment by pschatzmann** [comment]
You don't need the StreamCopyT. Just read the bytes from i2s like you would do it e.g. with files:
```
int16_t iBuffer[FLASH_RECORD_SIZE];
size_t bytes_read = i2s.readBytes((uint8_t*)iBuffer, FLASH_RECORD_SIZE*2); 
```
Please note that you get an array with in16_t values, so you will need to convert it to floats yourself. Audio float samples are usually scaled so that they are in the range of -1.0 and +1.0.
